zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
+
+    // Queued Blocks
+    //
+    /// A queue of unverified blocks.
+    ///
+    /// Contains a list of unverified blocks at each block height. In most cases,
+    /// the checkpoint verifier will store zero or one block at each height.
+    ///
+    /// Blocks are verified in order, when there is a chain from a subsequent
+    /// checkpoint, back to the previous checkpoint. Each checkpoint range is
+    /// used to verify all the blocks between the previous and target
+    /// checkpoints.
+    ///
+    /// The first checkpoint does not have any ancestors, so it only verifies the
+    /// genesis block.
+    queued: BTreeMap<BlockHeight, QueuedBlockList>,
+
+    /// The current progress of this verifier.
+    verifier_progress: Progress<BlockHeight>,
+}
+
+/// The CheckpointVerifier implementation.
+///
+/// Contains non-service utility functions for CheckpointVerifiers.
+impl CheckpointVerifier {
+    /// Return a checkpoint verification service, using the provided `checkpoint_list`.
+    ///
+    /// This function should be called only once for a particular checkpoint list (and
+    /// network), rather than constructing multiple verification services based on the
+    /// same checkpoint list. To Clone a CheckpointVerifier, you might need to wrap it
+    /// in a `tower::Buffer` service.
+    //
+    // Currently only used in tests.
+    //
+    // We'll use this function in the overall verifier, which will split blocks
+    // between BlockVerifier and CheckpointVerifier.
+    #[cfg(test)]
+    fn new(
+        checkpoint_list: impl IntoIterator<Item = (BlockHeight, BlockHeaderHash)>,
+    ) -> Result<Self, Error> {
+        Ok(CheckpointVerifier {
+            checkpoint_list: CheckpointList::new(checkpoint_list)?,
+            queued: <BTreeMap<BlockHeight, QueuedBlockList>>::new(),
+            // We start by verifying the genesis block, by itself
+            verifier_progress: Progress::BeforeGenesis,
+        })
+    }
+
+    /// Return the block height of the highest checkpoint in the checkpoint list.
+    ///
+    /// If there is only a single checkpoint, then the maximum height will be
+    /// zero. (The genesis block.)
+    ///
+    /// The maximum height is constant for each checkpoint list.
+    fn max_checkpoint_height(&self) -> BlockHeight {
+        self.checkpoint_list.max_height()
+    }
+
+    /// Return the current verifier's progress.
+    ///
+    /// If verification has not started yet, returns `BeforeGenesis`.
+    ///
+    /// If verification is ongoing, returns `PreviousCheckpoint(height)`.
+    /// `height` increases as checkpoints are verified.
+    ///
+    /// If verification has finished, returns `FinalCheckpoint`.
+    fn previous_checkpoint_height(&self) -> Progress<BlockHeight> {
+        self.verifier_progress
+    }
+
+    /// Return the start of the current checkpoint range.
+    ///
+    /// Returns None if verification has finished.
+    fn current_start_bound(&self) -> Option<Bound<BlockHeight>> {
```

Comment:
Not really, the exact bound depends on whether we verify a checkpoint range using `(Exclusive(_), Inclusive(_))` or `(Inclusive(_), Exclusive(_))`. Both are valid choices, the first requires special handling for the genesis block, the second requires special handling for the final checkpoint.

The choice of bounds is an implementation detail of `CheckpointVerifier`, because it needs to be kept consistent with `process_checkpoint_range()`, and the other functions it calls.